blackjax/tempered_smc.py

```diff
@@ -0,0 +1,251 @@
+    new_mcmc_state: Callable
+        How to create a new MCMC state for the MCMC routine to start
+    resampling_method: Callable
+        A random function that resamples generated particles based of weights
+    mcmc_stop_criterion: Callable
+        A function that takes all the MCMC states, the mcmc iteration count and returns False when the chain should stop
+
+    Returns
+    -------
+    A callable that takes a rng_key and a Pytree that contains the current state
+    of the chain and that returns a new state of the chain along with
+    information about the transition.
+    """
+
+    def one_step(rng_key: jnp.ndarray, state: SMCState, delta: float) -> Tuple[SMCState, SMCInfo]:
+        """
+
+        Parameters
+        ----------
+        rng_key: DeviceArray[int],
+            JAX PRNGKey for randomness
+        state: SMCState
+            Current state of the tempered SMC algorithm
+        delta: float
+            Increment for `state.lamda
+
+        Returns
+        -------
+        state: SMCState,
+            The new state of the tempered SMC algorithm
+         info: SMCInfo,
+            Additional information on the SMC step
+        """
+        n_particles = jax.tree_flatten(state.particles)[0][0].shape[0]
+
+        tempered_potential_fn = lambda pytree: -state.lamda * potential_fn(pytree)
+
+        step_mcmc_kernel = jax.vmap(mcmc_kernel_factory(tempered_potential_fn), in_axes=(0, 0))
+        rng_key, resampling_key = jax.random.split(rng_key, 2)
+
+        stop_criterion = lambda carry: mcmc_stop_criterion(carry[0], carry[1])
+
+        def mcmc_loop_body(carry):
+            i, mcmc_state, key = carry
+            key, body_key = jax.random.split(key, 2)
+            op_keys = jax.random.split(body_key, n_particles)
+            mcmc_state, _ = step_mcmc_kernel(op_keys, mcmc_state)
+            return i + 1, mcmc_state, key
+
+        initial_mcmc_states = jax.vmap(new_mcmc_state, in_axes=[0, None])(state.particles, tempered_potential_fn)
+        total_steps, last_mcmc_state, _ = jax.lax.while_loop(stop_criterion,
+                                                             mcmc_loop_body,
+                                                             (0, initial_mcmc_states, rng_key))
+        particles = last_mcmc_state.position
+
+        log_weights = -delta * jax.vmap(potential_fn, in_axes=[0])(particles)
+        weights = _normalize(log_weights)
+        resampling_index = resampling_method(weights, resampling_key)
+        particles = jax.tree_map(lambda x: x[resampling_index], particles)
+
+        state = SMCState(particles, state.lamda + delta)
+        info = SMCInfo(weights, total_steps)
+        return state, info
+
+    return one_step
+
+
+def _normalize(log_weights):
+    """ Normalize log-weights into weights
+    """
+    w = jnp.exp(log_weights - jnp.max(log_weights))
+    return w / w.sum()
+
+
+def _log_ess(log_weights):
```

Comment:
Everything ess-related could go in a `inference/smc/ess.py`

Reply:
ok